docs/architecture/adr-008-app-caller-cbs.md

```diff
@@ -0,0 +1,249 @@
+
+This requires a second layer of callbacks. The IBC application already gets the result of the packet from core IBC, but currently there is no standardized way to pass this information on to an actor module/smart contract.
+
+## Definitions
+
+- Actor: an actor is an on-chain module (this may be a hardcoded module in the chain binary or a smart contract) that wishes to execute custom logic whenever IBC receives packet flow that it has either sent or received. It **must** be addressable by a string value.
+
+## Decision
+
+Create a standardized callback interface that actors can implement. IBC applications (or middleware that wraps IBC applications) can now call this callback to route the result of the packet/channel handshake from core IBC to the IBC application to the original actor on the sending chain. IBC applications can route the packet receipt to the destination actor on the receiving chain.
+
+IBC actors may implement the following interface:
+
+```go
+type IBCActor interface {
+    // OnChannelOpen will be called on the IBCActor when the channel opens
+    // this will happen either on ChanOpenAck or ChanOpenConfirm
+    OnChannelOpen(ctx sdk.Context, portID, channelID, version string)
+
+    // OnChannelClose will be called on the IBCActor if the channel closes
+    // this will be called on either ChanCloseInit or ChanCloseConfirm and if the channel handshake fails on our end
+    // NOTE: currently the channel does not automatically close if the counterparty fails the handhshake so actors must be prepared for an OpenInit to never return a callback for the time being
+    OnChannelClose(ctx sdk.Context, portID, channelID string)
+
+    // OnRecvPacket will be called on the IBCActor after the IBC Application
+    // handles the RecvPacket callback if the packet has an IBC Actor as a receiver.
+    OnRecvPacket(ctx sdk.Context, packet channeltypes.Packet, relayer string)
+
+    // OnAcknowledgementPacket will be called on the IBC Actor
+    // after the IBC Application handles its own OnAcknowledgementPacket callback
+    OnAcknowledgmentPacket(
+        ctx sdk.Context,
+        packet channeltypes.Packet,
+        ack exported.Acknowledgement,
+        relayer string
+    )
+
+    // OnTimeoutPacket will be called on the IBC Actor
+    // after the IBC Application handles its own OnTimeoutPacket callback
+    OnTimeoutPacket(
+        ctx sdk.Context,
+        packet channeltypes.Packet,
+        relayer string
+    )
+}
+```
+
+The Packet interface will get extended to add `GetSender` and `GetReceiver` methods. These may return an address
+or they may return the empty string. The address may reference an IBCActor or it may be a regular user address. Any 
+IBC application or middleware that uses these methods must handle these cases.
+
+```go
+type Packet interface {
+    // existing Packet methods
+
+    // may return the empty string
+    GetSender() string
```

Comment:
Where does the non-empty value come from?

Reply:
Wouldn't the non-empty value come from the address that sent the IBC message, and the address that's going to receive the IBC message?